### PR TITLE
Make clippy a git subtree instead of a git submodule

### PR DESCRIPTION
**Pull request by oli-obk**
r? @eddyb 

cc #70651 

documentation at https://github.com/rust-lang/rust/pull/70654

### CODE REVIEW

**Comment by eddyb** [comment]
Whoa, clippy commit hashes were kept intact, this makes me *really* happy!11

**Comment by Mark-Simulacrum** [comment]
bootstrap changes look good (presuming they work, though the mingw-check job should confirm that clippy check works -- and it seems to have passed).

**Comment by flip1995** [comment]
Does this also enable running (at least UI) tests for Clippy on every PR?

**Comment by oli-obk** [comment]
> Does this also enable running (at least UI) tests for Clippy on every PR?

yes. We already run these right now on every PR, but we used to ignore it if clippy failed to successfully run its tests.

**Comment by Dylan-DPC** [comment]
Removed Prioritization label as this doesn't require it (based on our discussion in zulip)

**Comment by spastorino** [comment]
@Mark-Simulacrum you asked me to ping you here, this is giving conflicts so cc @oli-obk

**Comment by oli-obk** [comment]
Note: this is going to regress with the next clippy bump

**Comment by Mark-Simulacrum** [comment]
Okay, obviously this is ~impossible to review, but I checked over it a bit locally and it looks good.

So, let's do this!

@bors r+ p=100 (we're gating on clippy as of this PR, so high priority)

**Comment by oli-obk** [comment]
> Okay, obviously this is ~impossible to review, but I checked over it a bit locally and it looks good.

Only

* a182450 
* 1ef5a93 
* 06c4481 

are actual changes to this repo, the rest is just clippy being moved into the repo.

**Comment by Mark-Simulacrum** [comment]
Yep, I reviewed those three manually.

**Comment by oli-obk** [comment]
Thanks for the quick work! I'm very excited to be trying this out for clippy.

**Comment by flip1995** [comment]
CI logs show, that all Clippy tests passed. 

Also, currently the clippy fmt test is run on CI. We might want to turn this of, until clippy uses the same rustfmt version + `rustfmt.toml` as rustc. But this has to be done in Clippy.

**Comment by Mark-Simulacrum** [comment]
Yeah not sure what's up with that. We'll want to remove clippy from toolstate at least for sure since it should always be test-pass now (I suspect the above might be spurious? we're no longer recording clippy toolstate I think as of this PR)

**Comment by Manishearth** [comment]
So is this also gating on subtree clippy? (Not sure if we were doing that in two steps) @oli-obk do we have a set of instructions for what we should do about syncs?

**Comment by mati865** [comment]
Instructions were added in https://github.com/rust-lang/rust/pull/70654

**Comment by flip1995** [comment]
> So is this also gating on subtree clippy?

If I understand the changes correctly, Clippy now gets checked like rustdoc. So yes, this also gates at subtree clippy.

**Comment by matthiaskrgr** [comment]
> Instructions were added in #70654
Should we have a copy of that inside the clippy repo as well?

**Comment by flip1995** [comment]
We should change the documentation in clippy that talks about updating clippy in rust-lang/rust (I think that exists?)

**Comment by oli-obk** [comment]
The reason toolstate fails is that we now don't report any toolstate for clippy anymore, and thus toolstate assumes "failure". I completely forgot about that part of toolstate, I'll look into it tomorrow

File: src/bootstrap/test.rs [hunk 1]
```diff
@@ -548,9 +548,7 @@ impl Step for Clippy {
 
             builder.add_rustc_lib_path(compiler, &mut cargo);
 
-            if try_run(builder, &mut cargo.into()) {
-                builder.save_toolstate("clippy-driver", ToolState::TestPass);
-            }
```
Comment (flip1995) [inline src/bootstrap/test.rs:553]: This porbably caused rust-highfive to think that clippy is test-fail now.

**Comment by petrochenkov** [comment]
This will break rebase on master for all people in the near future:
```
error: The following untracked working tree files would be overwritten by checkout:
        src/tools/clippy/.editorconfig
        src/tools/clippy/.gitattributes
        src/tools/clippy/.github/ISSUE_TEMPLATE.md
        src/tools/clippy/.github/PULL_REQUEST_TEMPLATE.md
        src/tools/clippy/.github/deploy.sh
        ...
```

The solution is to get rid of the clippy submodule before the rebase:
```
git submodule deinit src/tools/clippy
```

**Pull request merged**